Enable pg_nccl.reduce_scatter to perform vector ReduceScatter for uneven input splits

aashaka opened the pull request:
Summary: A vector reduce_scatter requires each process to reduce and scatter an input tensor according to the input list provided. Internally, pg_nccl.reduce_scatter will coalesce a list of pg_nccl._reduce_oop to implement a vector reduce-scatter in the case when the any input shape is different in the input list. Otherwise, it will perform a ncclReduceScatter as usual.

- This change adds a `CoalescedWorkNCCL` class which encapsulates the WorkNCCL requests from coalesced operations. A `.wait()` on a CoalescedWorkNCCL request will call a wait on each of the WorkNCCL requests that are coalesced.

- This change adds an out-of-place `_reduce_oop` function to ProcessGroupNCCL. It allows reducing an input tensor and placing the output in a separate output tensor. Since reduce_scatter provides an out-of-place API, a reduce_scatter_v semantic implemented inside `pg_nccl.reduce_scatter` also needs to support out-of-place, for which an out-of-place reduce is required to be added.

Test Plan: Added a new test `test_reduce_scatter_v_cuda` for reduce_scatter_v to `distributed_nccl_spawn`.

Differential Revision: D38478781

kwen2501 commented on torch/_C/_distributed_c10d.pyi:
```diff
@@ -209,6 +209,21 @@ class ProcessGroup:
+        root: int,
+        op=ReduceOp.SUM,
+    ) -> Work: ...
+    @overload
```
**For your use case**, do you need both API signatures or just need one? I tend to think just need the Tensor version at the Python level.
Also, for clarity of the API name, you can name it `_reduce_oop`.

aashaka replied:
I have removed the API signature that is not needed and renamed `_reduce` to `_reduce_oop` everywhere.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -16,6 +16,8 @@ std::string opTypeToString(OpType opType) {
     case OpType::REDUCE:
       return "REDUCE";
+    case OpType::_REDUCE:
+      return "_REDUCE";
```
Same here, can add OOP at the end.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroup.hpp:
```diff
@@ -48,6 +48,7 @@ enum class OpType : std::uint8_t {
   RECVANYSOURCE = 14,
   BARRIER = 15,
   _REDUCE_SCATTER_BASE = 16,
+  _REDUCE = 17,
```
Same here.

kwen2501 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1090,6 +1090,39 @@ that adds a prefix to each key inserted to the store.
+              py::arg("root"),
+              py::arg("op") = ::c10d::ReduceOp::SUM,
+              py::call_guard<py::gil_scoped_release>())
+
```
Same here, do you need two signatures?

kwen2501 commented:
Maybe also worth writing a bit in the PR description about the context --
You need out-of-place reduce from the backend because you want to compose a reduce_scatter_v pattern at the Python front end using coalesced reduces. Today, the `dist.reduce_scatter` API supports out-of-place, so, if the reduce_scatter_v pattern is implemented under the `dist.reduce_scatter` API, you would need out-of-place support as well.

kwen2501 commented:
Also good to comment the above in code.

Commit:
Enable pg_nccl.reduce_scatter to perform vector ReduceScatter for uneven input splits (#82924)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/82924

A vector reduce_scatter requires each process to reduce and scatter an input tensor according to the input list provided.
Internally, pg_nccl.reduce_scatter will coalesce a list of pg_nccl._reduce_oop to implement a vector reduce-scatter in the case when the any input shape is different in the input list. Otherwise, it will perform a ncclReduceScatter as usual.

- This change adds a `CoalescedWorkNCCL` class which encapsulates the WorkNCCL requests from coalesced operations. A `.wait()` on a CoalescedWorkNCCL request will call a wait on each of the WorkNCCL requests that are coalesced.

- This change adds an out-of-place `_reduce_oop` function to ProcessGroupNCCL. It allows reducing an input tensor and placing the output in a separate output tensor. Since reduce_scatter provides an out-of-place API, a reduce_scatter_v semantic implemented inside `pg_nccl.reduce_scatter` also needs to support out-of-place, for which an out-of-place reduce is required to be added.

Test Plan: Added a new test `test_reduce_scatter_v_cuda` for reduce_scatter_v to `distributed_nccl_spawn`.

Differential Revision: D38478781

fbshipit-source-id: 0157acdee4e9a1dd328a27d4e30c3b81c4523039

kwen2501 reviewed (approved):
LGTM.

kwen2501 commented:
@pytorchbot merge

kwen2501 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2202,6 +2202,46 @@ def test_reduce_sum_cuda_twice(self):
+                if async_val:
+                    req.wait()
+
+                expected_value = 2 + (10 * (len(group) - 1))
```
nit: use variable defined above instead of instant numbers.

kwen2501 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2202,6 +2202,46 @@ def test_reduce_sum_cuda_twice(self):
+            end_len = start_len + input_split_sizes[rank]
+            sum_len = sum(input_split_sizes)
+            master_value = 2
+            worker_value = 10
```
nit:
rename `master_value` --> `value_to_self`
rename `worker_value` --> `value_to_others`
to be clearer
